### PR TITLE
 [sitecore-jss-nextjs]: Redirects with double and more parameters has been fixed for netlify #SXA-7554

### PR DESCRIPTION
**Pull request by sc-ruslanmatkovskyi**
- [x] This PR follows the [Contribution Guide](https://github.com/Sitecore/jss/blob/dev/CONTRIBUTING.md)
- [x] Changelog updated
- [ ] Upgrade guide entry added

## Description / Motivation

This PR includes a fix for an issue with Netlify, where URL parameters are being sorted by alphabet.
This causes redirects with two or more query parameters to not match the expected patterns. 
To resolve this, a new method has been added that generates all possible permutations of the query parameters based on their arrangement. This ensures that the redirect logic can correctly handle unordered query strings, allowing for proper matching with URL patterns, regardless of how parameters are ordered in the incoming requests.

## Testing Details

- [x] Unit Test Added
- [x] Manual Test/Other (Please elaborate)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts [hunk 1]
```diff
@@ -192,22 +192,26 @@ export class RedirectsMiddleware extends MiddlewareBase {
...
+      ? modifyRedirects.find((redirect: RedirectInfo & { matchedQueryString?: string }) => {
           redirect.pattern = redirect.pattern.replace(RegExp(`^[^]?/${language}/`, 'gi'), '');
           redirect.pattern = `/^\/${redirect.pattern
             .replace(/^\/|\/$/g, '')
             .replace(/^\^\/|\/\$$/g, '')
             .replace(/^\^|\$$/g, '')
             .replace(/(?<!\\)\?/g, '\\?')
-            .replace(/\$\/gi$/g, '')}[\/]?$/gi`;
+            .replace(/\$\/gi$/g, '')}[\/]?$/i`;
+          const matchedQueryString = this.isPermutedQueryMatch({
```
Comment (yavorsk) [inline packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts:203]: can you add a comment here explaining why we use this permute function? to be clear for anyone looking at the code in the future.

File: packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts [hunk 2]
```diff
@@ -285,4 +289,54 @@ export class RedirectsMiddleware extends MiddlewareBase {
...
     return redirect;
   }
+
+  /**
+   * Generates all possible permutations of an array of key-value pairs.
+   * This is used to create every possible combination of URL query parameters.
+   * @param {Array<[string, string]>} array - The array of key-value pairs to permute.
+   * @returns {Array<Array<[string, string]>>} - A 2D array where each inner array is a unique permutation of the input.
+   */
+  private getPermutations(array: [string, string][]): [string, string][][] {
```
Comment (yavorsk) [inline packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts:299]: I think it makes sense to move this in the base sitecore-jss package. in src/utils/utils.ts .What do you think?

**Review by yavorsk — CHANGES_REQUESTED** [review]
Looks good :) see few comments

File: packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts (hunk 1 above)
Comment (sc-ruslanmatkovskyi) [reply packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts:203]: Thanks! I've added the comments

File: packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts (hunk 2 above)
Comment (sc-ruslanmatkovskyi) [reply packages/sitecore-jss-nextjs/src/middleware/redirects-middleware.ts:299]: Moved

**Review by yavorsk — APPROVED** [review]
👍

**Pull request merged**